### PR TITLE
fix bfill, ffill and pad when calling with df.interpolate with column…

### LINKED ISSUE #12918: DataFrame.ffill behaves different than DataFrame.interpolate(method='ffill') along axes
It looks like `df.ffill(axis=0)` has the same behavior as `test_df.interpolate(method='ffill', axis=1)`.

```
from pandas.util.testing import assert_frame_equal
import numpy as np
import pandas as pd

n = np.nan
test_df = pd.DataFrame([[0, 2, n, n],
                        [1, n, 4, 6],
                        [n, 3, 5, n]])

assert_frame_equal(
    test_df.interpolate(method='ffill', axis=1),
    test_df.ffill(axis=0),
)
```

Is this the desired behavior?

```
In [2]: pandas.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Darwin
OS-release: 15.3.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.1
setuptools: 20.3.1
Cython: 0.23.4
numpy: 1.11.0
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.1.2
sphinx: 1.3.5
patsy: 0.4.0
dateutil: 2.5.2
pytz: 2016.3
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.5
matplotlib: 1.4.3
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.6.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
jinja2: 2.7.3
boto: 2.39.0
```

### LINKED ISSUE #29146: df.interpolate(method='pad')  axis is not consistent with df.fillna(method='pad')
```df.interpolate(method='pad')``` behaves weirdly regarding the axis. 

```python
df = pd.DataFrame([[1,np.nan,3],[np.nan, 2,np.nan],[2,3,np.nan]])

     0    1    2
0  1.0  NaN  3.0
1  NaN  2.0  NaN
2  2.0  3.0  NaN


df.fillna(method='pad', axis=0)

  0    1    2
0  1.0  NaN  3.0
1  1.0  2.0  3.0
2  2.0  3.0  3.0

df.interpolate(method='pad', axis=0)

   0    1    2
0  1.0  1.0  3.0
1  NaN  2.0  2.0
2  2.0  3.0  3.0

df.interpolate(method='linear', axis=0)
     0    1    2
0  1.0  NaN  3.0
1  1.5  2.0  3.0
2  2.0  3.0  3.0
```

Besides, ```df.interpolate(method='pad')``` does not seem to respond to `limit_area`

```python
df.interpolate(method='pad',limit_area='inside',axis=1)
     0    1    2
0  1.0  NaN  3.0
1  1.0  2.0  3.0
2  2.0  3.0  3.0

df.interpolate(method='linear',limit_area='inside', axis=0)

     0    1    2
0  1.0  NaN  3.0
1  1.5  2.0  NaN
2  2.0  3.0  NaN
```

#### Output of ``pd.show_versions()``

<details>

pandas           : 0.25.2
numpy            : 1.16.5
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.3.1
setuptools       : 41.4.0
Cython           : None
pytest           : 5.0.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.8.0
pandas_datareader: 0.8.0
bs4              : None
bottleneck       : None
fastparquet      : 0.3.0
gcsfs            : 0.3.1
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 0.13.0
pytables         : None
s3fs             : 0.3.5
scipy            : 1.3.1
sqlalchemy       : None
tables           : None
xarray           : 0.14.0
xlrd             : None
xlwt             : None
xlsxwriter       : None
</details>

### LINKED ISSUE #33956: BUG: throwing error in interpolate depending on dtype of column names
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas. (on 862db6421256, last commit where build works as of 20:42 2020-05-03).

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
df
Out[6]: 
     A     B     C
0  1.0   2.0   3.0
1  2.0   4.0   6.0
2  3.0   6.0   9.0
3  4.0   NaN   NaN
4  NaN   8.0   NaN
5  5.0  10.0  30.0
df.interpolate(method='ffill', axis=1)
```
Throws an error `ValueError: Index column must be numeric or datetime type when using ffill method other than linear. Try setting a numeric or datetime index column before interpolating.` But the following code works:

```python
df.columns = [1, 2, 3]
df
Out[9]: 
     1     2     3
0  1.0   2.0   3.0
1  2.0   4.0   6.0
2  3.0   6.0   9.0
3  4.0   NaN   NaN
4  NaN   8.0   NaN
5  5.0  10.0  30.0
df.interpolate(method='ffill', axis=1)
Out[10]: 
     1     2     3
0  1.0   2.0   3.0
1  2.0   4.0   6.0
2  3.0   6.0   9.0
3  4.0   6.0   9.0
4  4.0   8.0   9.0
5  5.0  10.0  30.0
```
#### Problem description

Throwing an error should not depend on the dtype of column names.

#### Expected Output

```python
    A     B     C
0  1.0   2.0   3.0
1  2.0   4.0   6.0
2  3.0   6.0   9.0
3  4.0   6.0   9.0
4  4.0   8.0   9.0
5  5.0  10.0  30.0
```

#### Output of ``pd.show_versions()``

```
pd.show_versions()
INSTALLED VERSIONS
------------------
commit           : 862db6421256cb7a00ae3e88a4a6999347b76271
python           : 3.8.2.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.3.0-51-generic
Version          : #44~18.04.2-Ubuntu SMP Thu Apr 23 14:27:18 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.1.0.dev0+1463.g862db6421
numpy            : 1.18.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1
setuptools       : 46.1.3.post20200325
Cython           : 0.29.17
pytest           : 5.4.1
hypothesis       : 5.10.4
sphinx           : 3.0.3
blosc            : None
feather          : None
xlsxwriter       : 1.2.8
lxml.etree       : 4.5.0
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.13.0
pandas_datareader: None
bs4              : 4.9.0
bottleneck       : 1.3.2
fastparquet      : 0.3.3
gcsfs            : None
matplotlib       : 3.2.1
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : 0.17.0
pytables         : None
pyxlsb           : None
s3fs             : 0.4.2
scipy            : 1.4.1
sqlalchemy       : 1.3.16
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.15.1
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.48.0
```

</details>

### PR DESCRIPTION
**Pull request by CloseChoice**
… dtype string (#33956)

- [x] closes #33956 
- [x] closes #12918 
- [x] closes #29146 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @CloseChoice for the PR.

not sure we should be adding tests for incorrect/undoumented behaviour. see https://github.com/pandas-dev/pandas/issues/33956#issuecomment-627313044

**Comment by simonjayhawkins** [comment]
xref #12918 maybe the tests could test for equality with either DataFrame.ffill or DataFrame.fillna(method='ffill) and update the DataFrame.interpolate docstring, see https://github.com/pandas-dev/pandas/issues/12918#issuecomment-211424101 and https://github.com/pandas-dev/pandas/issues/12918#issuecomment-419425660

**Comment by jreback** [comment]
can you merge master

**Comment by CloseChoice** [comment]
> xref #12918 maybe the tests could test for equality with either DataFrame.ffill or DataFrame.fillna(method='ffill) and update the DataFrame.interpolate docstring, see [#12918 (comment)](https://github.com/pandas-dev/pandas/issues/12918#issuecomment-211424101) and [#12918 (comment)](https://github.com/pandas-dev/pandas/issues/12918#issuecomment-419425660)

I went to the interpolate function back and forth and I come to the conclusion that we totally should fix this first. When we want to ensure that `fillna(method='ffill')` and `interpolate(method='ffill')` behave the same way, why shouldn't we just call `fillna` inside of `interpolate` in this case?

I am going to work on a PR to fix this problem, am curious for your feedback.

File: pandas/core/generic.py [hunk 1]
```diff
@@ -6881,30 +6881,41 @@ def interpolate(
...
 
         axis = self._get_axis_number(axis)
+        index = self._get_axis(axis)
+
+        if isinstance(self.index, MultiIndex) and method != "linear":
+            raise ValueError(
+                "Only `method=linear` interpolation is supported on MultiIndexes."
+            )
+
+        if method in ["backfill", "bfill", "pad", "ffill"]:
```
Comment (simonjayhawkins) [inline pandas/core/generic.py:6891]: can you add a comment here that for these methods, limit_direction and limit_area are being ignored and include a link to https://github.com/pandas-dev/pandas/issues/26796

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @CloseChoice for the updates, is it feasible for the fixes for #12918 and #29146 be split off into a pre-cursor PR. smaller PRs with more atomic changes are easier to review/quicker to merge. also needs a whatsnew.

**Comment by CloseChoice** [comment]
> Thanks @CloseChoice for the updates, is it feasible for the fixes for #12918 and #29146 be split off into a pre-cursor PR. smaller PRs with more atomic changes are easier to review/quicker to merge. also needs a whatsnew.

Thanks for your review. Actually fixing #12918 and #29146 has the side effect of fixing #33956, since the `fillna`-methods don't have the issue described in #33956. So I don't see possibility to split this into two PRs though I totally agree with you to have small, atomic PRs.

File: pandas/core/generic.py (hunk 1 above)
Comment (CloseChoice) [reply pandas/core/generic.py:6893]: done

File: pandas/tests/frame/methods/test_interpolate.py [hunk 2]
```diff
@@ -296,3 +297,45 @@ def test_interp_string_axis(self, axis_name, axis_number):
         result = df.interpolate(method="linear", axis=axis_name)
         expected = df.interpolate(method="linear", axis=axis_number)
         tm.assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("axis", [0, 1])
+    def test_interp_ffill(self, axis):
```
[the remark this answers is not in the document]
Comment (CloseChoice) [reply pandas/tests/frame/methods/test_interpolate.py]: done

File: pandas/core/generic.py [hunk 3]
```diff
@@ -6881,30 +6881,43 @@ def interpolate(
...
+            return self.fillna(
+                method=method,
+                axis=axis,
+                inplace=inplace,
+                limit=limit,
+                downcast=downcast,
+            )
 
+        # todo: change interpolation so that no transposing is necessary
```
Comment (jreback) [inline pandas/core/generic.py:6902]: TODO

File: pandas/tests/frame/methods/test_interpolate.py [hunk 4]
```diff
@@ -202,7 +202,8 @@ def test_interp_leading_nans(self, check_scipy):
             result = df.interpolate(method="polynomial", order=1)
             tm.assert_frame_equal(result, expected)
 
-    def test_interp_raise_on_only_mixed(self):
+    @pytest.mark.parametrize("axis", [0, 1])
```
Comment (jreback) [inline pandas/tests/frame/methods/test_interpolate.py:205]: use the axis fixture (e.g. just pass axis as the arg)

File: pandas/tests/frame/methods/test_interpolate.py [hunk 5]
```diff
@@ -213,7 +214,7 @@ def test_interp_raise_on_only_mixed(self):
             }
         )
         with pytest.raises(TypeError):
-            df.interpolate(axis=1)
+            df.astype("object").interpolate(axis=axis)
```
Comment (jreback) [inline pandas/tests/frame/methods/test_interpolate.py:217]: why is this changed?

Comment (jreback) [reply pandas/tests/frame/methods/test_interpolate.py:217]: check the error message here

File: pandas/tests/frame/methods/test_interpolate.py [hunk 6]
```diff
@@ -296,3 +297,18 @@ def test_interp_string_axis(self, axis_name, axis_number):
         result = df.interpolate(method="linear", axis=axis_name)
         expected = df.interpolate(method="linear", axis=axis_number)
         tm.assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("method", ["ffill", "bfill", "pad"])
+    @pytest.mark.parametrize("axis", [0, 1])
```
Comment (jreback) [inline pandas/tests/frame/methods/test_interpolate.py:302]: axis fixture

File: pandas/core/generic.py [hunk 7]
```diff
@@ -6881,30 +6881,43 @@ def interpolate(
...
         axis = self._get_axis_number(axis)
+        index = self._get_axis(axis)
+
+        if isinstance(self.index, MultiIndex) and method != "linear":
+            raise ValueError(
+                "Only `method=linear` interpolation is supported on MultiIndexes."
+            )
+
+        # for the methods backfill, bfill, pad, ffill limit_direction and limit_area
+        # are being ignored, see #26796 for more information
```
Comment (jreback) [inline pandas/core/generic.py:6892]: list as gh-26796

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_interpolate.py [hunk 8]
```diff
@@ -296,3 +297,18 @@ def test_interp_string_axis(self, axis_name, axis_number):
...
+    def test_interp_fillna_methods(self, axis, method):
+        # GH 33956
```
Comment (jreback) [inline pandas/tests/frame/methods/test_interpolate.py:304]: can you put the actual issue number here.

File: doc/source/whatsnew/v1.1.0.rst [hunk 9]
```diff
@@ -337,6 +337,7 @@ Backwards incompatible API changes
 - Combining a ``Categorical`` with integer categories and which contains missing values
   with a float dtype column in operations such as :func:`concat` or :meth:`~DataFrame.append`
   will now result in a float column instead of an object dtyped column (:issue:`33607`)
+- :meth:`DataFrame.interpolate` uses the correct axis convention. Previously interpolating along columns lead to interpolation along indices and vice versa.
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:340]: list the closed issues.

move this to the missing bug fix section.

this note is note indicative of what is actually being fixed, can you be more descriptive.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.1.0.rst (hunk 9 above)
Comment (CloseChoice) [reply doc/source/whatsnew/v1.1.0.rst]: done

File: pandas/tests/frame/methods/test_interpolate.py (hunk 5 above)
Comment (CloseChoice) [reply pandas/tests/frame/methods/test_interpolate.py:221]: done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @CloseChoice very nice
